File: scripts/sounds.py
# scripts/sounds.py
# Система звуков и музыки: загрузка, проигрывание, управление

import pygame
import os
import random

# Инициализация микшера (если ещё не сделано в main.py)
if not pygame.mixer.get_init():
    pygame.mixer.init(frequency=44100, size=-16, channels=2, buffer=512)

class SoundManager:
    def __init__(self):
        self.sounds = {}          # словарь звуковых эффектов
        self.music = None         # текущая музыка
        self.music_volume = 0.5
        self.effects_volume = 0.8
        self.muted = False

        # Путь к папке звуков
        self.base_path = "../assets/sounds/"

        # Загружаем звуки по умолчанию (если файлов нет — пропускаем)
        self.load_default_sounds()

    def load_default_sounds(self):
        """Загружает типичные звуки (можно расширять)"""
        sound_files = {
            "jump": "jump.wav",
            "step": ["step1.wav", "step2.wav", "step3.wav"],  # несколько для вариации
            "collect": "collect.wav",
            "hurt": "hurt.wav",
            "bonus": "bonus.wav",
            "perk_unlock": "perk.wav",
            "death": "death.wav",
            "win": "win.wav",
            "click": "click.wav"
        }

        for name, file_or_list in sound_files.items():
            if isinstance(file_or_list, list):
                self.sounds[name] = []
                for f in file_or_list:
                    path = os.path.join(self.base_path, f)
                    if os.path.exists(path):
                        self.sounds[name].append(pygame.mixer.Sound(path))
            else:
                path = os.path.join(self.base_path, file_or_list)
                if os.path.exists(path):
                    self.sounds[name] = pygame.mixer.Sound(path)

    def play(self, sound_name, loops=0, volume=None):
        """Проиграть звуковой эффект"""
        if self.muted:
            return

        if sound_name in self.sounds:
            sound = self.sounds[sound_name]
            if isinstance(sound, list):
                sound = random.choice(sound)  # случайный из вариантов шагов и т.д.
            if volume is not None:
                old_vol = sound.get_volume()
                sound.set_volume(volume)
                sound.play(loops)
                sound.set_volume(old_vol)  # возвращаем старый
            else:
                sound.set_volume(self.effects_volume)
                sound.play(loops)

    def play_music(self, music_file, loops=-1, fade_ms=2000):
        """Проиграть фоновую музыку"""
        path = os.path.join(self.base_path, music_file)
        if os.path.exists(path):
            pygame.mixer.music.load(path)
            pygame.mixer.music.set_volume(self.music_volume)
            pygame.mixer.music.play(loops, fade_ms=fade_ms)
            self.music = music_file

    def stop_music(self, fade_ms=1000):
        pygame.mixer.music.fadeout(fade_ms)
        self.music = None

    def pause_music(self):
        pygame.mixer.music.pause()

    def unpause_music(self):
        pygame.mixer.music.unpause()

    def set_music_volume(self, vol):
        self.music_volume = max(0.0, min(1.0, vol))
        pygame.mixer.music.set_volume(self.music_volume)

    def set_effects_volume(self, vol):
        self.effects_volume = max(0.0, min(1.0, vol))
        # Можно обновить все загруженные звуки, но проще менять при play

    def mute_all(self):
        self.muted = True
        pygame.mixer.music.set_volume(0)

    def unmute_all(self):
        self.muted = False
        pygame.mixer.music.set_volume(self.music_volume)

    def toggle_mute(self):
        if self.muted:
            self.unmute_all()
        else:
            self.mute_all()

# Глобальный менеджер звуков (создавай один экземпляр)
sound_manager = SoundManager()

File: js/sounds.js
// js/sounds.js — версия без скачивания, всё по прямым ссылкам, CORS-safe ресурсы

class SoundManager {
  constructor() {
    this.sounds = {};
    this.currentMusic = null;
    this.volumeEffects = 0.6;
    this.volumeMusic = 0.3;
    this.muted = false;
    this.userInteracted = false;

    // Прямые ссылки на звуки с CORS-разрешением (public domain / CC0)
    this.loadRemoteSounds();
  }

  loadRemoteSounds() {
    // Надёжные источники без CORS-блоков
    const soundMap = {
      jump: 'https://assets.codepen.io/21542/jump.mp3',
      step: 'https://assets.codepen.io/21542/step.mp3',
      collect: 'https://assets.codepen.io/21542/coin.mp3',
      hurt: 'https://assets.codepen.io/21542/hit.mp3',
      bonus: 'https://assets.codepen.io/21542/powerup.mp3',
      perk_unlock: 'https://assets.codepen.io/21542/levelup.mp3',
      explosion: 'https://assets.codepen.io/21542/explosion.mp3',
      win: 'https://assets.codepen.io/21542/victory.mp3',
      death: 'https://assets.codepen.io/21542/death.mp3',
      click: 'https://assets.codepen.io/21542/click.mp3',
      level_up: 'https://assets.codepen.io/21542/levelup.mp3'
    };

    for (const [name, url] of Object.entries(soundMap)) {
      const audio = new Audio(url);
      audio.preload = 'auto';
      this.sounds[name] = audio;
    }

    // Музыка — один трек с CORS-разрешением
    this.musicUrl = 'https://assets.codepen.io/21542/8bit-dungeon.mp3';
  }

  play(name, vol = 1.0) {
    if (this.muted || !this.userInteracted) return;
    const sound = this.sounds[name];
    if (sound) {
      sound.volume = this.volumeEffects * vol;
      sound.currentTime = 0;
      sound.play().catch(() => console.log('Звук не проигрался, ждём клика'));
    }
  }

  playRandomStep() {
    this.play('step', 0.4);
  }

  playMusic() {
    if (this.muted || !this.userInteracted || this.currentMusic) return;

    const music = new Audio(this.musicUrl);
    music.loop = true;
    music.volume = this.volumeMusic;
    music.play().catch(() => console.log('Музыка ждёт взаимодействия'));
    this.currentMusic = music;
  }

  stopMusic() {
    if (this.currentMusic) {
      this.currentMusic.pause();
      this.currentMusic = null;
    }
  }

  toggleMute() {
    this.muted = !this.muted;
    if (this.currentMusic) this.currentMusic.volume = this.muted ? 0 : this.volumeMusic;
  }
}

const soundManager = new SoundManager();

// Включаем звук после любого клика или нажатия
window.addEventListener('click', () => { soundManager.userInteracted = true; soundManager.playMusic(); }, {once: true});
window.addEventListener('keydown', () => { soundManager.userInteracted = true; soundManager.playMusic(); }, {once: true});
